src/CodeGen_Hexagon.cpp

```diff
@@ -175,6 +150,62 @@ Stmt sloppy_unpredicate_loads(Stmt s) {
     return SloppyUnpredicateLoads().mutate(s);
 }
 
+class InjectHVXLocks : public IRMutator2 {
+public:
+    InjectHVXLocks(const Target &t) : target(t) {}
+    bool uses_hvx = false;
+private:
+    using IRMutator2::visit;
+
+    Stmt visit(const For *op) {
+        if (op->for_type == ForType::Parallel) {
+            bool old_uses_hvx = uses_hvx;
+            uses_hvx = false;
+            Stmt body = mutate(op->body);
+            if (uses_hvx) {
+                body = acquire_hvx_context(body, target);
+            }
+            uses_hvx = old_uses_hvx;
+            Stmt new_for = For::make(op->name, op->min, op->extent, op->for_type,
+                             op->device_api, body);
+            if (old_uses_hvx) {
+                Stmt call_hvx_lock = call_halide_qurt_hvx_lock(target);
```

Comment:
Hmm, it looks like this doesn't unlock HVX prior to entering the for loop, which I didn't think about before. Is this safe? We're relying on the same thread calling qurt_hvx_lock twice in a row to be OK. Needs a comment if it's safe, if not, we need to figure out how to unlock prior to the for loop. This is tricky due to the destructor implementation.

Reply:
Well, I see now below that we just unconditionally unlock HVX in the thread pool. How about just doing that here instead (unconditionally call qurt_hvx_unlock prior to entering the parallel for loop)? And with a comment explaining why it's safe.

Reply:
Ok I was doing things wrong here. I want to acquire the lock after the parallel loop (because the parallel loop will have released the lock in its destructor block) instead I was acquiring it before. I have fixed this now and pushed up changes that incorporate other comments in your review. I am testing these changes as we speak but feel free to review.